kernel/userspace.c

```diff
@@ -475,6 +475,114 @@ void _k_object_uninit(void *object)
 	ko->flags &= ~K_OBJ_FLAG_INITIALIZED;
 }
 
+/*
+ * Copy to/from helper functions used in syscall handlers
+ */
+void *z_user_alloc_from_copy(void *src, size_t size)
+{
+	void *dst = NULL;
+	int key;
+
+	key = irq_lock();
+
+	/* Does the caller in user mode have access to read this memory? */
+	if (Z_SYSCALL_MEMORY_READ(src, size)) {
+		goto out_err;
+	}
+
+	dst = z_thread_malloc(size);
+	if (!dst) {
+		printk("out of thread resource pool memory (%zu)", size);
+		goto out_err;
+	}
+
+	memcpy(dst, src, size);
+out_err:
+	irq_unlock(key);
+	return dst;
+}
+
+static int user_copy(void *dst, void *src, size_t size, bool to_user)
+{
+	int ret = EFAULT;
+	int key;
+
+	key = irq_lock();
+
+	/* Does the caller in user mode have access to this memory? */
+	if (to_user ? Z_SYSCALL_MEMORY_WRITE(dst, size) :
+	              Z_SYSCALL_MEMORY_READ(src, size)) {
+		goto out_err;
+	}
+
+	memcpy(dst, src, size);
+	ret = 0;
+out_err:
+	irq_unlock(key);
+	return ret;
+}
+
+int z_user_from_copy(void *dst, void *src, size_t size)
+{
+	return user_copy(dst, src, size, false);
+}
+
+int z_user_to_copy(void *dst, void *src, size_t size)
+{
+	return user_copy(dst, src, size, true);
+}
+
+char *z_user_string_alloc_copy(char *src, size_t maxlen)
+{
+	size_t actual_len;
+	int key, err;
+	char *ret = NULL;
+
+	key = irq_lock();
+	actual_len = z_user_string_nlen(src, maxlen, &err);
```

Comment:
Could we just use retval < 0 as an error indication?

Reply:
I wanted to keep this the same as strnlen which returns a size_t.
Is there some advantage to this I am missing?

Reply:
Only simpler use and easier error dispatching (but this is my personal opinion).
```
ssize_t size = z_user_string_nlen(src, maxlen);
if (size < 0) {
      return -EFAULT;
}
```